src/cascadia/TerminalSettingsModel/VsSetupConfiguration.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "VsSetupConfiguration.h"
+
+using namespace Microsoft::Terminal::Settings::Model;
+
+std::vector<VsSetupConfiguration::VsSetupInstance> VsSetupConfiguration::QueryInstances()
+{
+    std::vector<VsSetupConfiguration::VsSetupInstance> instances;
+
+    // SetupConfiguration is only registered if Visual Studio is installed
+    ComPtrSetupQuery pQuery{ wil::CoCreateInstanceNoThrow<SetupConfiguration, ISetupConfiguration2>() };
+    if (pQuery == nullptr)
+    {
+        return instances;
+    }
+
+    // Enumerate all valid instances of Visual Studio
+    wil::com_ptr<IEnumSetupInstances> e;
+    THROW_IF_FAILED(pQuery->EnumInstances(&e));
+
+    std::array<ComPtrSetupInstance, 1> rgpInstance;
+    auto result = e->Next(1, &rgpInstance[0], NULL);
+    while (S_OK == result)
+    {
+        // wrap the COM pointers in a friendly interface
+        instances.emplace_back(VsSetupConfiguration::VsSetupInstance{ pQuery, rgpInstance[0] });
+        result = e->Next(1, &rgpInstance[0], NULL);
+    }
+
+    THROW_IF_FAILED(result);
+
+    return instances;
+}
+
+/// <summary>
+/// Takes a relative path under a Visual Studio installation and returns the absolute path.
+/// </summary>
+std::wstring VsSetupConfiguration::ResolvePath(ComPtrSetupInstance pInst, std::wstring_view relativePath)

```

Comment:
minor nit about these -- since the ComPtr argument is taken by copy, each of these causes an AddRef/Release (presumably, unless the compiler is smart?) on its argument for every call. It may not be necessary if the caller is holding an owning reference.

I believe the idiomatic way to take non-owning parameters is via raw pointer arguments (ala `xxxSetupInstance*`).

A less-idiomatic way, which requires the holder to have a `wil::com_ptr` themselves, is to slap a `const &` on it. 😁